// rbs34/utils.js
const { actions } = require("openhab")

let logger = log("rbs34.utils");
const SECRETS_DIR = "/openhab/conf/secrets/";


class TaggedRuleName {
    constructor(tag) {this.tag = tag.toUpperCase()};
    get(name) {return "[" + this.tag + "] " + name};
};


function sendEventGhostCommand(command) {
    const err_msg = "Unable to send command to EventGhost: {}"
    if (items.getItem("desktopComputerOnline").state != "ON") {
        logger.error(err_msg, "Computer Offline");
    }
    logger.info("Sending command to Event Ghost: {}", command);
    const url = "http://vweber-desk:2187/?rbs34." + command;
    actions.HTTP.sendHttpGetRequest(url, 60000);
    return true
};


class JoinPushBody {
    constructor(device) {
        const joinSecrets = require(SECRETS_DIR + "join.json");
        if (!joinSecrets.hasOwnAttribute(device)) {
            logger.error("Could not fetch deviceId: {}", device);
            return;
        }
        this.apikey = joinSecrets.apikey;
        this.deviceId = joinSecrets[deviceId];
    }
}


module.exports = {
    TaggedRuleName,
    sendEventGhostCommand,
    SECRETS_DIR,
    JoinPushBody
};
